runtime/src/transaction_batch.rs

```diff
@@ -55,6 +55,24 @@ impl<'a, 'b> Drop for TransactionBatch<'a, 'b> {
     }
 }
 
+pub trait IntoCowForSanitizedTransaction<'a>:
+    'a + Clone + std::borrow::Borrow<SanitizedTransaction>
+{
+    fn into_cow(self) -> Cow<'a, [SanitizedTransaction]>;
+}
+
+impl<'a> IntoCowForSanitizedTransaction<'a> for SanitizedTransaction {
+    fn into_cow(self) -> Cow<'a, [SanitizedTransaction]> {
+        Cow::Owned(vec![self])
+    }
+}
+
+impl<'a> IntoCowForSanitizedTransaction<'a> for &'a SanitizedTransaction {
+    fn into_cow(self) -> Cow<'a, [SanitizedTransaction]> {
+        Cow::Borrowed(std::slice::from_ref(self))
+    }
+}
```

Comment:
There is a single call to this function.
And the amount of code it is hiding is trivial.
I also do not fully understand if there is a higher level meaning to this transformation, than just hiding the `Cow` constructor call.
If there is none, it might be only adding a burden for the reader, by hiding useful details, rather than simplifying the call site.

Reply:
> I also do not fully understand if there is a higher level meaning to this transformation, than just hiding the Cow constructor call.

there's no meaning in this case other than hiding the Cow constructor call, while i indeed didn't write higher level meanings at other places as you know... ;)

> If there is none, it might be only adding a burden for the reader, by hiding useful details, rather than simplifying the call site with sensible degree of abstraction.

hmm, i think using existential type like this at fn defitnition actually simplify things:

```rust
pub fn prepare_sanitized_batch_without_locking<'a>(
    &'a self,
    transaction: impl IntoCowForSanitizedTransaction<'a>,
) -> TransactionBatch<'a, 'a> {
    ...
```

at call site

```rust
let batch = bank.prepare_sanitized_batch_without_locking(transaction);
```

Reply:
Just FYI, you can use `Cow` directly in the function API.  It is part of the standard library, so you can assume everyone knows what it is:

```rust
pub fn prepare_sanitized_batch_without_locking<'a, Tx>(
    &'a self,
    transaction: Cow<'a, [Tx]>,
) -> TransactionBatch<'a, 'a>
where
  Tx: Clone + Borrow<SanitizedTransaction>,
{
    ...
```

But you might be right that `IntoCowForSanitizedTransaction` could make the function signature shorter.
I personally might prefer to see the `Clone`, `Borrow<SanitizedTransaction>` and the slice directly in the function API, but I also tend to write rather complex constraints.

Also, I wonder if the `Tx` argument could be simplified.
`Borrow` is a pretty generic constraint that is more common in libraries, rather than in the application level code.

Reply:
fyi, after some research, it turned out this can greatly be simplified...: https://github.com/solana-labs/solana/pull/32304

Reply:
> Just FYI, you can use `Cow` directly in the function API. It is part of the standard library, so you can assume everyone knows what it is:
> ...
> I personally might prefer to see the Clone, Borrow<SanitizedTransaction> and the slice directly in the function API, but I also tend to write rather complex constraints.

my personal preference is opposite: I prefer to hide types as much as possible, maybe so that we can switch to `beef` or something else easily. i regard these as implementation details to be hidden.

> ```rust
>     ...
>     transaction: Cow<'a, [Tx]>,
>     ...
> ```

given #32304, this is off-topi but i'd rather avoid taking slice here because we want to create a batch from a single transaction. i think that single part is kind of important as  this api's invariant. both call-sites (rpc simulator, new schduler) assumes only about single transaction and providing _flexibility_ of multiple transaction is rather unsafe.

> `Borrow` is a pretty generic constraint that is more common in libraries, rather than in the application level code.

yeah, but i think it's understandable that helper code (this trait) could tend to be like written in libraries coding style even in the application level code.

Reply:
> my personal preference is opposite: I prefer to hide types as much as possible, maybe so that we can switch to `beef` or something else easily. i regard these as implementation details to be hidden.

I agree that implementation details should generally better be hidden.
But certain details are exposed due to the mechanics of an API, rather than semantics.
For example, in Rust, the ownership model is explicit, and so `&` and `&mut` are part of an API that users are expecting to see.
And similarly there are `Arc`, and `Pin` that clarify the mechanics of how the value is used, rather than talk about the semantics of the interaction with it or some internal implementation.
In my mind, `Cow` is in the same class.  It is a type that exists only to help work with the ownership and would not exist, should a garbage collector be managing the object lifetimes.
As such, it is something that the users want to see, and changing it to anything else will require a matching change to the code using the API anyway.

> > ```rust
> >     ...
> >     transaction: Cow<'a, [Tx]>,
> >     ...
> > ```
> 
> given #32304, this is off-topi but i'd rather avoid taking slice here because we want to create a batch from a single transaction. i think that single part is kind of important as this api's invariant. both call-sites (rpc simulator, new schduler) assumes only about single transaction and providing _flexibility_ of multiple transaction is rather unsafe.

I just exposed the type that was already there, when I was writing the example.
Must admit, I do not fully understand the semantics of this slice =)